crates/futures/src/polyfill.rs

```diff
@@ -0,0 +1,208 @@
+use wasm_bindgen::JsCast;
+use web_sys::{MessageEvent, Worker};
+
+macro_rules! console_log {
+    ($($t:tt)*) => (log(&format_args!($($t)*).to_string()))
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    fn log(s: &str);
+}
+
+const DEFAULT_TIMEOUT: f64 = 10.0;
+
+const HELPER_CODE: &'static str = "
+onmessage = function (ev) {
+    try {
+        switch (ev.data[0]) {
+            case 'wait': {
+                let [_, ia, index, value, timeout] = ev.data;
+                console.log('wait event inside a worker');
+                let result = Atomics.wait(ia, index, value, timeout);
+                postMessage(['ok', result]);
+                break;
+            }
+            default: {
+                throw new Error('Wrong message sent to wait helper: ' + ev.data.join(','));
+            }
+        }
+    } catch (e) {
+        console.log('Exception in wait helper');
+        postMessage(['error', 'Exception']);
+    }
+}
+";
+
+thread_local! {
+    static HELPERS: RefCell<Vec<Rc<Worker>>> = RefCell::new(vec![]);
+}
+
+fn alloc_helper() -> Rc<Worker> {
+    HELPERS.with(|helpers| {
+        if let Some(helper) = helpers.borrow_mut().pop() {
+            return helper;
+        }
+
+        let mut initialization_string = "data:application/javascript,".to_owned();
+        let encoded: String = encode_uri_component(HELPER_CODE).into();
+        initialization_string.push_str(&encoded);
+
+        return Rc::new(Worker::new(&initialization_string).expect("Should create a Worker"));
+    })
+}
+
+fn free_helper(helper: &Rc<Worker>) {
+    HELPERS.with(move |helpers| {
+        helpers.borrow_mut().push(helper.clone());
+    });
+}
+
+pub fn wait_async(index: u32, value: i32) -> Result<Promise, JsValue> {
+    wait_async_with_timeout(index, value, DEFAULT_TIMEOUT)
+}
+
+fn get_array_item(array: &JsValue, index: u32) -> JsValue {
+    Reflect::get(array, &JsValue::from(index))
+        .expect(&format!("Array should contain the index {}", index))
+}
+
+// Atomics.waitAsync always returns a promise. Throws standard errors
+// for parameter validation.  The promise is resolved with a string as from
+// Atomics.wait, or, in the case something went completely wrong, it is
+// rejected with an error string.
+pub fn wait_async_with_timeout(index: u32, value: i32, timeout: f64) -> Result<Promise, JsValue> {
+    let memory_buffer = wasm_bindgen::memory()
+        .dyn_into::<WebAssembly::Memory>()
+        .expect("Should cast a memory to WebAssembly::Memory")
+        .buffer();
+
+    let indexed_array = Int32Array::new(&memory_buffer);
+
+    if !indexed_array.buffer().has_type::<SharedArrayBuffer>() {
+        console_log!("polyfill, not a SharedArrayBuffer");
+        return Err(Error::new("Indexed array must be created from SharedArrayBuffer").into());
+    }
+
+    // Optimization, avoid the helper thread in this common case.
+    if Atomics::load(&indexed_array, index)? != value {
```

Comment:
If `index` is changed to `&AtomicI32`, you don't need to call a JS API here and can instead simply do `ptr.load(SeqCst)`